lib/private/Security/Signature/SignatureManager.php

```diff
@@ -0,0 +1,823 @@
+		} catch (SignatoryNotFoundException|SignatureException) {
+			try {
+				$signatory = $this->getSafeRemoteSignatory($signatoryManager, $signedRequest);
+			} catch (SignatoryNotFoundException $e) {
+				$this->manageDeprecatedSignatory($knownSignatory);
+				throw $e;
+			}
+
+			$signedRequest->setSignatory($signatory);
+			$this->storeSignatory($signatory);
+			$this->verifySignedRequest($signedRequest);
+		}
+	}
+
+
+	/**
+	 * @param ISignatoryManager $signatoryManager
+	 * @param IIncomingSignedRequest $signedRequest
+	 *
+	 * @return ISignatory
+	 * @throws InvalidKeyOriginException
+	 * @throws SignatoryNotFoundException
+	 */
+	private function getSafeRemoteSignatory(
+		ISignatoryManager $signatoryManager,
+		IIncomingSignedRequest $signedRequest,
+	): ISignatory {
+		$signatory = $signatoryManager->getRemoteSignatory($signedRequest);
+		if ($signatory === null) {
+			throw new SignatoryNotFoundException('empty result from getRemoteSignatory');
+		}
+		if ($signatory->getKeyId() !== $signedRequest->getKeyId()) {
+			throw new InvalidKeyOriginException('keyId from signatory not related to the one from request');
+		}
+
+		return $signatory->setProviderId($signatoryManager->getProviderId());
+	}
+
+	private function setOutgoingSignatureHeader(
+		IOutgoingSignedRequest $signedRequest,
+		string $method,
+		string $path,
+		string $dateHeader,
+	): void {
+		$header = [
+			'(request-target)' => $method . ' ' . $path,
+			'content-length' => strlen($signedRequest->getBody()),
+			'date' => gmdate($dateHeader),
+			'digest' => $signedRequest->getDigest(),
+			'host' => $signedRequest->getHost()
+		];
+
+		$signedRequest->setSignatureHeader($header);
+	}
+
+
+	/**
+	 * @param IOutgoingSignedRequest $signedRequest
+	 */
+	private function setOutgoingClearSignature(IOutgoingSignedRequest $signedRequest): void {
+		$signing = [];
+		$header = $signedRequest->getSignatureHeader();
+		foreach (array_keys($header) as $element) {
+			$value = $header[$element];
+			$signing[] = $element . ': ' . $value;
+			if ($element !== '(request-target)') {
+				$signedRequest->addHeader($element, $value);
+			}
+		}
+
+		$signedRequest->setClearSignature(implode("\n", $signing));
+	}
+
+
+	private function setOutgoingSignedSignature(IOutgoingSignedRequest $signedRequest): void {
+		$clear = $signedRequest->getClearSignature();
+		$signed = $this->signString(
+			$clear, $signedRequest->getSignatory()->getPrivateKey(), $signedRequest->getAlgorithm()
+		);
+		$signedRequest->setSignedSignature($signed);
+	}
+
+	private function signingOutgoingRequest(IOutgoingSignedRequest $signedRequest): void {
+		$signatureHeader = $signedRequest->getSignatureHeader();
+		$headers = array_diff(array_keys($signatureHeader), ['(request-target)']);
+		$signatory = $signedRequest->getSignatory();
+		$signatureElements = [
+			'keyId="' . $signatory->getKeyId() . '"',
+			'algorithm="' . $this->getChosenEncryption($signedRequest->getAlgorithm()) . '"',
```

Comment:
```suggestion
			'algorithm="' . $signedRequest->getAlgorithm() . '"',
```
Why not this directly?